lib/utils/compose.coffee

```diff
@@ -0,0 +1,746 @@
+		@_serviceToDataMap[service] = event
+
+	_handleInterrupt: =>
+		@_cancelled = true
+		@end()
+		process.exit(130) # 128 + SIGINT
+
+	start: =>
+		process.on('SIGINT', @_handleInterrupt)
+		@_tty.hideCursor()
+		@_services.forEach (service) =>
+			@streams[service].write({ status: 'Preparing...' })
+		@_runloop = createRunLoop(@_display)
+		@_startTime = Date.now()
+
+	end: (summary = null) =>
+		return if @_ended
+		@_ended = true
+		process.removeListener('SIGINT', @_handleInterrupt)
+		@_runloop.end()
+		@_runloop = null
+
+		@_clear()
+		@_renderStatus(true)
+		@_renderSummary(summary ? @_getServiceSummary())
+		@_tty.showCursor()
+
+	_display: =>
+		@_clear()
+		@_renderStatus()
+		@_renderSummary(@_getServiceSummary())
+		@_tty.cursorUp(@_services.length + 1) # for status line
+
+	_clear: ->
+		@_tty.deleteToEnd()
+		@_maxLineWidth = @_tty.currentWindowSize().width
+
+	_getServiceSummary: ->
+		_ = require('lodash')
+
+		services = @_services
+		serviceToDataMap = @_serviceToDataMap
+
+		_(services).map (service) ->
+			{ status, progress, error } = serviceToDataMap[service] ? {}
+			if error
+				return "#{error}"
+			else if progress
+				bar = renderProgressBar(progress, 20)
+				return "#{bar} #{status}" if status
+				return "#{bar}"
+			else if status
+				return "#{status}"
+			else
+				return 'Waiting...'
+		.map (data, index) ->
+			[ services[index], data ]
+		.fromPairs()
+		.value()
+
+	_renderStatus: (end = false) ->
+		moment = require('moment')
+		require('moment-duration-format')(moment)
+
+		@_tty.clearLine()
+		@_tty.write(@_prefix)
+		if end and @_cancelled
+			@_tty.writeLine('Build cancelled')
+		else if end
+			serviceCount = @_services.length
+			serviceStr = if serviceCount is 1 then '1 service' else "#{serviceCount} services"
+			runTime = Date.now() - @_startTime
+			durationStr = moment.duration(runTime // 1000, 'seconds').format()
+			@_tty.writeLine("Built #{serviceStr} in #{durationStr}")
+		else
+			@_tty.writeLine("Building services... #{@_spinner()}")
+
+	_renderSummary: (serviceToStrMap) ->
+		_ = require('lodash')
+		chalk = require('chalk')
+		truncate = require('cli-truncate')
+		strlen = require('string-width')
+
+		@_services.forEach (service, index) =>
+			str = _.padEnd(@_prefix + chalk.bold(service), @_prefixWidth)
+			str += serviceToStrMap[service]
+			if @_maxLineWidth?
+				str = truncate(str, @_maxLineWidth)
+			@_lineWidths[index] = strlen(str)
+
+			@_tty.clearLine()
+			@_tty.writeLine(str)
+
+class BuildProgressInline
+	constructor: (outStream, descriptors) ->
+		_ = require('lodash')
+		through = require('through2')
+
+		services = _.map(descriptors, 'serviceName')
+		eventHandler = @_renderEvent
+		streams = _(services).map (service) ->
+			stream = through.obj (event, _enc, cb) ->
+				eventHandler(service, event)
+				cb()
+			stream.pipe(outStream)
+			[ service, stream ]
+		.fromPairs()
+		.value()
+
+		offset = 10 # account for escape sequences inserted for colouring
```

Comment:
I'd much prefer this kind of stuff was added into the logging utils, rather than being done by hand at individual call sites.

Reply:
This has nothing to do with the logger. This is to correctly right-pad service names so their logs left-align.

Reply:
I know, but that feels like it should be part of the logger to me, rather than repeated several times where you're doing custom output